fletcher/base.py

```diff
@@ -673,29 +673,80 @@ def take(self, indices, allow_fill=False, fill_value=None):
         ValueError
             When `indices` contains negative values other than ``-1``
             and `allow_fill` is True.
-
         Notes
         -----
-        FletcherArray.take is called by ``Series.__getitem__``, ``.loc``,
+        ExtensionArray.take is called by ``Series.__getitem__``, ``.loc``,
         ``iloc``, when `indices` is a sequence of values. Additionally,
         it's called by :meth:`Series.reindex`, or any other method
         that causes realignemnt, with a `fill_value`.
 
+
         See Also
         --------
         numpy.take
         pandas.api.extensions.take
         """
-        from pandas.core.algorithms import take
+        threshold_ratio = 0.3
+
+        # this is the threshold to decide whether or not to concat everything first.
+        # Benchmarks were made on string, int32, int64, float32, float64 and it turns out that 0.3 is the value where it
+        # is best to switch to concatening everything first, both time-wise and memory-wise
+
+        length = len(self)
+        indices = np.asarray(indices, dtype=self._indices_dtype)
+        has_negative_indices = np.any(indices < 0)
+        allow_fill &= has_negative_indices
+        if allow_fill:
+            validate_indices(indices, length)
+        if (has_negative_indices and not allow_fill) or np.any(indices >= length):
+            # this will raise IndexError expected by pandas in all needed cases
+            indices = np.arange(length, dtype=self._indices_dtype).take(indices)
+        # here we guarantee that indices is numpy array of ints
+        # and we have checked that all indices are between -1/0 and len(self)
+
+        if not allow_fill:
+            if self.data.num_chunks == 1:
+                return FletcherArray(self.data.chunk(0).take(pa.array(indices)))
+
+            lengths = np.fromiter(map(len, self.data.chunks), dtype=np.int)
+            cum_lengths = lengths.cumsum()
+
+            bins = np.searchsorted(
+                cum_lengths, indices, side="right"
+            )  # one might want to parallel this operation
+
+            cum_lengths -= lengths
+            limits_idx = np.concatenate(
+                [[0], np.bincount(bins, minlength=self.data.num_chunks).cumsum()]
+            )
+
+            if pd.Series(bins).is_monotonic:
+                del bins
+                return self._take_on_chunks(
+                    indices, limits_idx=limits_idx, cum_lengths=cum_lengths
+                )
+            elif (
+                len(indices) / len(self) > threshold_ratio
+            ):  # check which method is going to take less memory
+                return self._take_on_concatenated_chunks(indices)
+            else:
+                sort_idx = get_group_index_sorter(bins, self.data.num_chunks)
+                del bins
+                indices = indices[sort_idx]
```

Comment:
to save some memory !
```suggestion
                indices = indices.take(sort_idx, out=indices)`
```